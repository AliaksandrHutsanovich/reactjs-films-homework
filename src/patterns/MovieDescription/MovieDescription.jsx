import React, { useMemo, memo } from 'react';
import PropTypes from 'prop-types';
import { Row, Col } from 'antd';

import '!style-loader!css-loader!sass-loader!antd/dist/antd.css'; //eslint-disable-line

import { Text } from '../../components';

const MovieDescription = ({
  title,
  genres,
  rating,
  titleFontType,
  ratingFontType,
}) => {
  const genresList = useMemo(
    () => (
      genres.reduce(
        (acc, genre) => (`${acc}, ${genre}`),
      )
    ),
    [genres], // additional memoization is possible
  );

  return (
    <Row justify="space-between">
      <Col span={18}>
        <Row>
          <Text size={18} type={titleFontType} strong>{title}</Text>
        </Row>
        <Row>
          <Text type="secondary">{genresList}</Text>
        </Row>
      </Col>
      <Col>
        <Text size={16} type={ratingFontType}>{rating}</Text>
      </Col>
    </Row>
  );
};

MovieDescription.propTypes = {
  title: PropTypes.string.isRequired,
  genres: PropTypes.arrayOf(
    PropTypes.string,
  ).isRequired,
  rating: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  titleFontType: PropTypes.oneOf([
    'normal',
    'white',
  ]),
  ratingFontType: PropTypes.oneOf([
    'white_with_background',
    'secondary_with_background',
  ]),
};

MovieDescription.defaultProps = {
  titleFontType: 'normal',
  ratingFontType: 'white_with_background',
};

export default memo(MovieDescription); // additional memoization is possible
