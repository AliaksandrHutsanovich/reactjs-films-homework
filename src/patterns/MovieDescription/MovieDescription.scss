.description {
  &__title {
    font: 15px HelveticaNeueBold;
  }

  &__text {
    font: 10px HelveticaNeueMedium;
    color: blue;
  }

  &__rating {
    font: 10px HelveticaNeueMedium;
    padding: 5px;
    color: blue;
    background-color: grey;
    border-radius: 1px;
  }
}
