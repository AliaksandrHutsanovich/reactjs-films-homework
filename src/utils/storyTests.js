const storyTests = (stories, renderer) => {
  Object.keys(stories).forEach((storyKey) => {
    if (typeof stories[storyKey] !== 'function') {
      return;
    }
    it(`${storyKey} should be rendered properly`, () => {
      renderer.render(stories[storyKey]());
      const result = renderer.getRenderOutput();

      expect(result).toMatchSnapshot();
    });
  });
};

export default storyTests;
