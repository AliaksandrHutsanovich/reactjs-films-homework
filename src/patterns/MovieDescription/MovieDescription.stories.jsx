import React from 'react';
import MovieDescription from './MovieDescription';

export default {
  component: MovieDescription,
  title: 'patterns/MovieDescription',
  decorators: [
    (story) => (
      <div
        style={{
          padding: '20px',
          width: '400px',
          height: '300px',
        }}
      >
        {story()}
      </div>
    ),
  ],
};

const genres = [
  'Action',
  'Adventure',
  'Fantasy',
];

export const cardMetaInfo = () => (
  <MovieDescription
    title="ASSASIN'S CREED"
    genres={genres}
    rating="4.7"
    ratingFontType="secondary_with_background"
  />
);
