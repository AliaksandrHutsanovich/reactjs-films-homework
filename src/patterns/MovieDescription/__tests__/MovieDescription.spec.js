import ShallowRenderer from 'react-test-renderer/shallow';
import storyTests from '../../../utils/storyTests';
import * as stories from '../MovieDescription.stories';

describe('<MovieDescription />', () => {
  const renderer = new ShallowRenderer();

  storyTests(stories, renderer);
});
