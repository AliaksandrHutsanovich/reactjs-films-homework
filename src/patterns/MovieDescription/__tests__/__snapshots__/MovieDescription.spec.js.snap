// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieDescription /> cardMetaInfo should be rendered properly 1`] = `
<ForwardRef
  justify="space-between"
>
  <ForwardRef
    span={18}
  >
    <ForwardRef>
      <UNDEFINED
        size={18}
        strong={true}
      >
        ASSASIN'S CREED
      </UNDEFINED>
    </ForwardRef>
    <ForwardRef>
      <UNDEFINED
        type="secondary"
      >
        Action, Adventure, Fantasy
      </UNDEFINED>
    </ForwardRef>
  </ForwardRef>
  <ForwardRef>
    <UNDEFINED
      size={16}
      type="secondary_with_background"
    >
      4.7
    </UNDEFINED>
  </ForwardRef>
</ForwardRef>
`;
